azure/Guidelines.md

```diff
@@ -84,84 +82,103 @@ https://<tenant-id>-<service-defined-root>.<service>.azure.net
-
-#### Existing property is removed
-
-If a property called `foo` that was present in v1 of the API needs to be removed, it must be done in a newer API version.
+### API Changes that require a version change
 
-#### New property added to response
+There are three groups of changes that may happen to an API.
 
-If a new property/field is added to the response of an API, the GET-PUT pipeline will be broken. Consider the case where a customer updates the value of a new property "A" from the Azure portal. Another customer does a GET of this resource using the SDK. The SDK will ignore the property since it does not understand it. From the SDK, the customer does a PUT using the model that was returned from the GET. This will overwrite the change made by the first customer from the portal.
+1. Changes made to an **EXISTING** API version due to security or compliance reasons.  We shall refer to these types of changes as _Compliance changes_.
+2. Changes made to an API that may cause a client making the API call to fail, such as removal of an endpoint or property or changing the format of the body.  We refer to these types of changes as _Breaking changes_.
+3. Additive changes made to an API that do not cause a client making the API call to fail, such as the addition of a new optional property or a new endpoint.  We refer to these types of changes as _Evolutionary changes_.
 
-#### New required property added to request
+With the exception of _Compliance changes_ (which are extremely rare), Azure services **MUST** update the version number of their API whenever there is a change to the API, no matter how small.  Customers will "lock the API version" so that their code does not fail when the service introduces new features.  They rely on the fact that an API version is a contract with the services that will never change.
 
-If a new property is made required in the request body, clients will have no way to set this and the request will fail.
+A _breaking change_ is any change in the API that may cause client or service code making the API call to fail. Obvious examples of such a change are the removal of an endpoint, adding or removing a required field or changing the format of the body (from XML to JSON for example). Even though we recommend clients ignore new fields, there are many libraries and clients that fail when new fields are introduced. Removing an endpoint from an API is always a _breaking change_.  Adding a new endpoint is always an _evolutionary change_.  Changes to properties may be _evolutionary_ or _breaking_ depending on the type of change and whether the change is to an input parameter or output parameter:
```

Comment:
@ravbhatnagar - could you clarify on what you mean by 'not properly handling new properties in the same api-version'?

To me it feels like there are two options for the RP implementation, with the following drawbacks:
1. **PUT-as-PATCH**: Template deployments become non-deterministic/repeatable.
2. **Strict PUT semantics**: An older version of the SDK can inadvertently perform a destructive update.

Reply:
There is a middle ground that is PUT as "replace everything the client should know about in the given API version and keep the rest as-is". Which is a variation of PUT-as-PATCH.

Reply:
Anthony - There are two options in my mind (and RP can chose) 
1. RP can explicitly disallow update to a resource which was created using v2 if the request comes from v1. Specially if the resource has properties set which are not exposed to v1. 400 bad request with a nice error message indicating that the resource was created with a new version, please move to latest version to update it. 
2. As Johan was saying, and this is deviating from what true PUT definition, but an option where I see a path to bringing some consistency in behavior across services. Else, in other options it will be next to impossible to get existing RPs change the behavior and fix PUTCH. So, yes, Microsoft Azure invents a new definition of PUT (on similar lines as Johan was saying). But atleast we have a path to achieve consistency across 200+ Azure services. Something to noodle on (there are definitely challenges/concerns I am aware of, but willing to explore further)